File: modules/libs/core/container/nouns_test.go
package container

import (
	"go/ast"
	"go/parser"
	"go/token"
	"io/fs"
	"path/filepath"
	"regexp"
	"slices"
	"strings"
	"testing"
)

// nouns are the words of ours ending in -ing or -ed that are ordinary English
// nouns, and what each one means.
//
// A type is a noun or a noun phrase: a gerund says what is happening to a
// thing rather than what the thing is, and a stranger meeting one cannot say
// what it holds or whose it is. A machine reading a name sees only how it
// ends, and no suffix tells a gerund from a word that merely finishes those
// letters — so the words that are not gerunds are written down with a gloss.
// It is a dictionary and not a list of exemptions: a reader applies the rule
// without reading it, and a word nothing is named by any more has to go.
var nouns = map[string]string{
	"embed":     "the scenario that gives a vault's chunks their vectors, in the family of verb-named use cases",
	"heading":   "a line a section of a note stands under",
	"indexing":  "the settings section about making a vault searchable, beside Recognition and Transcription",
	"rating":    "how well a card was answered",
	"reading":   "what a document was read as",
	"recording": "a sound file",
	"setting":   "one setting of the file",
}

// The rule is drawn at the exported names. An unexported type is met only
// inside the package that declares it, where the file and the package say what
// it is; an exported one travels, and is what a stranger meets cold.
var nounWords = regexp.MustCompile(`[A-Z]+(?:[a-z0-9]*)`)

// gerund is whether a name reads as a gerund or a participle: it ends in those
// letters and its last word is in no dictionary of ours.
func gerund(name string) bool {
	if !strings.HasSuffix(name, "ing") && !strings.HasSuffix(name, "ed") {
		return false
	}
	said := nounWords.FindAllString(name, -1)
	if len(said) == 0 {
		return true
	}
	_, known := nouns[strings.ToLower(said[len(said)-1])]
	return !known
}

// declared are the exported types one file declares.
func declared(file *ast.File) []string {
	var found []string
	ast.Inspect(file, func(node ast.Node) bool {
		spec, is := node.(*ast.TypeSpec)
		if is && spec.Name.IsExported() {
			found = append(found, spec.Name.Name)
		}
		return true
	})
	return found
}

// everyExportedType is every exported type of the core, against the file it
// stands in.
func everyExportedType(t *testing.T) map[string]string {
	t.Helper()
	found := map[string]string{}
	err := filepath.WalkDir("..", func(path string, entry fs.DirEntry, err error) error {
		if err != nil || entry.IsDir() || !strings.HasSuffix(path, ".go") {
			return err
		}
		if strings.HasSuffix(path, "_test.go") || strings.HasSuffix(path, ".pb.go") {
			return nil
		}
		file, err := parser.ParseFile(token.NewFileSet(), path, nil, parser.SkipObjectResolution)
		if err != nil {
			return err
		}
		for _, one := range declared(file) {
			found[one] = path
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	return found
}

// A type is a noun or a noun phrase. Verb phrases name functions and methods;
// nouns name the things they act on.
//
// One word came to name a flashcards pill, a notes result and a handler at
// once, and nobody could say which from the name. Plexing could not have been
// introduced without somebody writing the word into the dictionary above and
// being asked what it means.
func TestNoTypeOfTheCoreIsNamedByAGerundOrParticiple(t *testing.T) {
	found := everyExportedType(t)
	var wrong []string
	for name, path := range found {
		if gerund(name) {
			wrong = append(wrong, path+" declares "+name)
		}
	}
	slices.Sort(wrong)
	for _, one := range wrong {
		t.Error(one + ": a type is a noun or a noun phrase")
	}

	// A walk that read no type of the core is a rule checked against nothing,
	// and it passes. The count is a floor well under what the core declares.
	if len(found) < 150 {
		t.Fatalf("%d exported types read: the walk is not reading the core", len(found))
	}
}

// A word nothing is named by any more is a word nobody has to argue for, and a
// dictionary that keeps them fills up until it reads as a census. It only
// shrinks.
func TestEveryWordOfTheDictionaryNamesAType(t *testing.T) {
	said := map[string]bool{}
	for name := range everyExportedType(t) {
		if !strings.HasSuffix(name, "ing") && !strings.HasSuffix(name, "ed") {
			continue
		}
		if words := nounWords.FindAllString(name, -1); len(words) > 0 {
			said[strings.ToLower(words[len(words)-1])] = true
		}
	}
	var gone []string
	for word := range nouns {
		if !said[word] {
			gone = append(gone, word)
		}
	}
	slices.Sort(gone)
	for _, one := range gone {
		t.Error(one + " is in the dictionary and names no type: the dictionary only shrinks")
	}
}

// What the rule refuses, read against names written to be refused. The words
// it lets through are the dictionary's, and a name not ending in those letters
// is never this rule's business however it reads.
//
// Drawn is the participle the rule names and this reading cannot see: an
// irregular one has no ending to test for, and a machine that tried would be
// reading English rather than a suffix. A person catches those.
func TestWhatTheNounRuleRefuses(t *testing.T) {
	cases := []struct {
		says    string
		name    string
		allowed bool
	}{
		{"a gerund", "Plexing", false},
		{"a participle", "Configured", false},
		{"a gerund with a qualifier in front", "PlexFiling", false},
		{"a clause read as a name", "FolderMissing", false},
		{"a noun in the dictionary", "Heading", true},
		{"a compound of one", "OpenRecording", true},
		{"a plain noun", "Vault", true},
		{"a noun whose stem is a verb", "Editor", true},
		{"an irregular participle, which has no ending", "Drawn", true},
	}

	var refused, wanted []string
	for _, one := range cases {
		if gerund(one.name) {
			refused = append(refused, one.says)
		}
		if !one.allowed {
			wanted = append(wanted, one.says)
		}
	}
	if !slices.Equal(refused, wanted) {
		t.Errorf("the rule refuses %v, want %v", refused, wanted)
	}
}

File: modules/tools/lint/nouns.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'
import { declares, nouns, owed, refused, words } from './nouns.mjs'
import { sources } from './source.mjs'

const declared = () => {
  const found = []
  for (const { at, text } of sources(['.ts', '.vue'])) {
    for (const name of declares(text)) found.push({ at, name })
  }
  return found
}

/**
 * A gerund or a participle says what is happening to a thing rather than what
 * the thing is. `Plexing` could not have been introduced without somebody
 * writing the word into the dictionary below and being asked what it means.
 */
test('no type of the interface modules is named by a gerund or a participle', () => {
  const found = declared()
  const wrong = found
    .filter(({ name }) => refused(name) && !owed.includes(name))
    .map(({ at, name }) => `${at} declares ${name}`)
  assert.deepEqual(wrong, [])

  // A walk that read no declaration is a rule checked against nothing, and it
  // passes. A count alone cannot say which modules it read, so the furthest of
  // them is named: the phone is where a rule stops at a border first.
  assert.ok(found.length > 400, `${found.length} types read: the walk is not reading the modules`)
  assert.ok(
    found.some(({ at }) => at.endsWith('apps/mobile/src/core.ts')),
    "the walk did not read the phone's core.ts, so the rule stops at the mobile border",
  )
})

/**
 * A word nothing is named by any more is a word nobody has to argue for, and a
 * dictionary that keeps them fills up until it reads as a census. Both lists
 * only shrink.
 */
test('every word in the dictionary names something, and every debt is still owed', () => {
  const found = declared()
  const said = new Set()
  for (const { name } of found) {
    if (/(ing|ed)$/.test(name)) said.add(words(name).at(-1))
  }
  assert.deepEqual(
    Object.keys(nouns).filter((one) => !said.has(one)),
    [],
  )

  const names = new Set(found.map((one) => one.name))
  assert.deepEqual(
    owed.filter((one) => !names.has(one)),
    [],
  )
})

/**
 * What the rule refuses, read against names written to be refused. The words
 * it lets through are the dictionary's, and a name that does not end in those
 * letters is never this rule's business however it reads.
 */
test('what the noun rule refuses', () => {
  const cases = [
    { says: 'a gerund', allowed: false, name: 'Plexing' },
    { says: 'a participle', allowed: false, name: 'Configured' },
    { says: 'a gerund with a qualifier in front', allowed: false, name: 'PlexFiling' },
    { says: 'a noun in the dictionary', allowed: true, name: 'Heading' },
    { says: 'a compound of one', allowed: true, name: 'OpenRecording' },
    { says: 'a plain noun', allowed: true, name: 'Vault' },
    { says: 'a noun whose stem is a verb', allowed: true, name: 'Editor' },
    // Drawn is the participle the rule names and this walk cannot see: an
    // irregular one has no ending to test for, and a machine that tried would
    // be reading English rather than a suffix. A person catches those.
    { says: 'an irregular participle, which has no ending', allowed: true, name: 'Drawn' },
  ]

  const wrong = cases.filter((one) => refused(one.name)).map((one) => one.says)
  const wanted = cases.filter((one) => !one.allowed).map((one) => one.says)
  assert.deepEqual(wrong, wanted)
})

/** The declarations the walk has to find, and the clauses it must not read. */
test('what counts as a declaration', () => {
  const source = [
    'export interface Alpha { a: 1 }',
    'type Beta = string',
    'export type { Gamma } from "./elsewhere"',
    'import type { Delta } from "./elsewhere"',
    'export abstract class Epsilon {}',
    'export enum Zeta { One }',
    ' * type Eta is named only in a comment',
    '/*',
    'type Plexing = 1',
    '*/',
    'const theta = `',
    'type Filing = 2',
    '`',
  ].join('\n')
  assert.deepEqual(declares(source), ['Alpha', 'Beta', 'Epsilon', 'Zeta'])
})
